subsys/bluetooth/host/conn.c

```diff
@@ -903,8 +903,11 @@ void bt_conn_process_tx(struct bt_conn *conn)
 
 	LOG_DBG("conn %p", conn);
 
-	if (conn->state == BT_CONN_DISCONNECTED &&
-	    atomic_test_and_clear_bit(conn->flags, BT_CONN_CLEANUP)) {
+	bool disconnecting = conn->state == BT_CONN_DISCONNECTED ||
+			     conn->state == BT_CONN_DISCONNECTING ||
+			     conn->state == BT_CONN_DISCONNECT_COMPLETE;
```

Comment:
I'm wondering if it would make more sense to check for `conn->state != BT_CONN_CONNECTED`. 

Arguably we only want to process TX when in the connected state, right?

Reply:
This was also my initial idea, but frankly, the whole state machine in conn.c looks a little overcomplicated and didn't have time to analyze it deeply. On the other hand, you are correct, inside the send_buf() function there is a check to validate the state is CONNECTED.

Reply:
Is there any scenario where we are not in the `BT_CONN_CONNECTED` state and do not want to flush our TX queue?

Reply:
I will take a look, but probably this weekend, and I can redo tests next week to see if '!= CONNECTED' is OK.

Reply:
Hi @Thalley

I did some additional analysis and testing and this is what I found:

1. BT_CONN_CLEANUP flag is set by bt_conn_set_state() only when transition from BT_CONN_DISCONNECT_COMPLETE -> BT_CONN_DISCONNECTED. See this:

```
		/* Notify disconnection and queue a dummy buffer to wake
		 * up and stop the tx thread for states where it was
		 * running.
		 */
		switch (old_state) {
		case BT_CONN_DISCONNECT_COMPLETE:
			tx_notify(conn);

			/* Cancel Connection Update if it is pending */
			if ((conn->type == BT_CONN_TYPE_LE) &&
			    (k_work_delayable_busy_get(&conn->deferred_work) &
			     (K_WORK_QUEUED | K_WORK_DELAYED))) {
				k_work_cancel_delayable(&conn->deferred_work);
			}

			atomic_set_bit(conn->flags, BT_CONN_CLEANUP);
			k_poll_signal_raise(&conn_change, 0);
			/* The last ref will be dropped during cleanup */
			break;
```

so in bt_conn_process_tx(), the call to conn_cleanup() will be executed only when the state is BT_CONN_DISCONNECTED and BT_CONN_CLEANUP set. This means my change here is useless.

2. What fixes memory buffer release is the check for errors from send_buf(). If the connection state is different than BT_CONN_CONNECTED (tested in do_send_frag()) this function returns '-ENOTCONN'. And this is what should be handled.

3. What is still not clear to me is how other error codes should be treated. Like '-ENOBUFS' or '-ENOMEM'.

Reply:
- `-ENOBUFS` is only in the case of not enough "credits" for sending data to the controller. In that case, we still haven't popped `buf` from the tx queue, so no leaks. Unless the hci driver
- `-ENOMEM` is in the case where we don't have a buffer for allocating a fragment for sending parts of `buf`. In that case we also haven't popped `buf` from TX queue, so no leaks again
- Any other error codes from the controller [are converted to `-EIO`](https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/bluetooth/host/conn.c#L637) to avoid a scenario when the above cases are not true.